README.md
# Teapot

> Utilities for working with HTTP status codes, errors, and more.

[![npm version](https://badge.fury.io/js/node-teapot.svg)](https://badge.fury.io/js/node-teapot)
[![build status](https://travis-ci.org/greylocklabs/teapot.svg?branch=master)](https://travis-ci.org/greylocklabs/teapot)
[![coverage status](https://coveralls.io/repos/github/greylocklabs/teapot/badge.svg?branch=master)](https://coveralls.io/github/greylocklabs/teapot?branch=master)

---

Teapot is an HTTP utility library for JavaScript, which leverages the
[Node.js HTTP library](https://nodejs.org/api/http.html). It provides the following:

1. The ability to get an HTTP status code: `teapot.status(404)` and `teapot.status('not found')` would both
   return `404`.
2. Useful error classes to represent HTTP error codes:
    - `HTTPError`: Base class to represent an HTTP error
    - `ClientError` and `ServerError`: Classes to represent 4xx and 5xx errors
    - Classes for every unique HTTP error status code, ranging from `NotImplementedError` to `PaymentRequiredError`
3. A function to generate one of the HTTP error classes from a status code: `teapot.error(404)` would return an
   instance of `NotFoundError`. Great when handling responses from third-party APIs, when you might not know what
   status codes to expect all the time.

TypeScript definitions are included as well.

## Installation

With `yarn`:

```bash
$ yarn add node-teapot
```

With `npm`:

```bash
$ npm install node-teapot
```

## Usage

### Get a status code

There are a variety of ways to get a status code from a number or string message:

```js
teapot.status.code(404); // 404
teapot.status.code('not implemented'); // 405

teapot.status.codes['BAD GATEWAY']; // 502

teapot.status.MOVED_PERMANENTLY; // 301
```

### Get a canned status message

```js
teapot.status[200]; // "OK"
```

### Create an HTTP error

Teapot's errors are compatible with Koa and Express:

```js
throw new teapot.InternalServerError('Oops! Something went wrong.');
```

### Generate an error from a status code

```js
teapot.error(500) // returns instance of InternalServerError
teapot.error(204) // throws error because 204 is not an error code

teapot.error(404, 'My custom message', { // custom message w/ misc. additional properties
   expose: true,
   data: {
      misc: 'blah',
   },
})
```

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md) for guidelines.

## License

MIT License. See [LICENSE](LICENSE) file for details.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. The format is based on
[Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

---

## [Unreleased](https://github.com/greylocklabs/teapot/compare/3.1.1...HEAD)

- N/A

## [3.1.1](https://github.com/greylocklabs/teapot/compare/3.1.0...3.1.1) - 2020-02-24

### Changed

- Update dependencies, fix linting errors

## [3.1.0](https://github.com/greylocklabs/teapot/compare/3.0.5...3.1.0) - 2019-09-04

### Fixed

- TypeScript type definitions

### Changed

- Dependency upgrades
- Minor config file changes

## [3.0.5](https://github.com/greylocklabs/teapot/compare/3.0.4...3.0.5) - 2019-04-22

### Fixed

- Check for existence of `Error.captureStackTrace` before calling

## [3.0.4](https://github.com/greylocklabs/teapot/compare/3.0.2...3.0.4) - 2019-04-22

### Added

- Ability to set `data` property on errors, using `teapot.error` function.

## [3.0.2](https://github.com/greylocklabs/teapot/compare/3.0.1...3.0.2) - 2019-04-19

### Fixed

- Adds TypeScript declaration file

## [3.0.1](https://github.com/greylocklabs/teapot/compare/3.0.0...3.0.1) - 2019-04-19

### Fixed

- Correct package name in README + .yarnrc

## [3.0.0](https://github.com/greylocklabs/teapot/compare/2.0.1...3.0.0) - 2019-04-18

### Changed

- Moves to TypeScript
- New APIs - `teapot.status` function replaced with `teapot.status.code`
- Uses Jest for testing

## [2.0.1](https://github.com/greylocklabs/teapot/compare/1.2.0...2.0.1) - 2018-02-06

### Changed

A whole lot! The package has been renamed to `teapot` to make room for the new default export and avoid confusion
between it and Node's native `http` module. Some other changes:

- `createError` is now `teapot.error`; functionality is the same
- `status` is now `teapot.status`; functionality is the same
- Errors can be accessed via the default export `teapot.errors`, or by named import
- The package bundle now only includes the `dist` folder to shrink size
- Minor changes to examples
- Far more extensive test suite

## [1.2.0](https://github.com/greylocklabs/teapot/compare/1.1.0...1.2.0) - 2018-02-01

### Added

- A new `createError` function! It lets you take a status code and create the proper `ClientError`
  or `ServerError` that goes with it. Examples:
  - `const err = clientError(404, 'My not found message');`
- Update dependencies
- Examples updated to use `createError`
- More tests

### Changed

- Classes that inherit `ClientError` or `ServerError` now have a static `code` method to get the
  HTTP status code associated with the error. For example, `NotFoundError.code() === 404`

## [1.1.0](https://github.com/greylocklabs/teapot/compare/1.0.1...1.1.0) - 2018-01-17

### Changed

- Switch to Node.js Markdown linting tool
- Update packages

## [1.0.1](https://github.com/greylocklabs/teapot/compare/1.0.0...1.0.1) - 2017-11-18

### Changed

- Minor changes to NPM scripts

### Removed

- No longer generating JSDoc files, using Doclets instead

## [1.0.0](https://github.com/greylocklabs/teapot/releases/tag/1.0.0) - 2017-11-17

### Added

- Status code and error utility modules
- Tests for all modules
- Full code coverage
- JSDoc documentation
- Examples using Koa and Express


src/status/index.js
/**
 * @file Exports status() function and other useful utilities and constants
 * @module status
 *
 * @author Ty-Lucas Kelley <ty@greylocklabs.com> (https://greylocklabs.com)
 * @copyright Copyright (c) 2017-2018 Greylock Labs. See LICENSE file for details.
 */

import http from 'http';

/**
 * Fills in the status.codes array, and allows for things like status[404] and status['not found']
 * @private
 *
 * @param {Function} func - The status function to add properties to
 *
 * @returns {number[]} Array of status codes, in order
 */
const addStatusCodesAndMessages = (func) => {
    const codes = http.STATUS_CODES;
    const list = [];

    Object.keys(codes).forEach((code) => {
        const msg = codes[code];
        const num = Number.parseInt(code);

        func[msg] = num;
        func[msg.toLowerCase()] = num;
        func[msg.toUpperCase()] = num;
        func[num] = msg;
        func[code] = msg;

        list.push(num);
    });

    return list;
};

/**
 * Takes in an HTTP status code or its associated message and returns the code
 * @public
 *
 * @param {string|number} input - The status code or message, i.e. 404, '404', 'not found', or 'Not Found'
 *
 * @returns {number} HTTP status code
 *
 * @throws {Error} Invalid status code or message is provided
 * @throws {TypeError} Input is not a number or string
 */
const status = (input) => {
    if (typeof input === 'number') {
        if (status[input]) {
            return input;
        }

        throw new Error(`Invalid status code ${input}`);
    } else if (typeof input === 'string') {
        if (isNaN(input)) { // eslint-disable-line no-restricted-globals
            const code = status[input.toLowerCase()];

            if (code) {
                return code;
            }

            throw new Error(`Invalid status message ${input}`);
        } else if (Number(input) % 1 !== 0) {
            throw new Error(`Invalid status code ${input}; must be an integer`);
        }

        return status(Number.parseInt(input));
    }

    throw new TypeError('Input must be a number or string');
};

/**
 * Check if a given status code is a 1xx information response
 * @public
 *
 * @param {string|number} input - Status code
 *
 * @returns {boolean} True if code is 1xx
 */
status.isInfo = (input) => status.codes.includes(Number.parseInt(input)) && String(input).charAt(0) === '1';

/**
 * Check if a given status code is a 2xx success response
 * @public
 *
 * @param {string|number} input - Status code
 *
 * @returns {boolean} True if code is 2xx
 */
status.isSuccess = (input) => status.codes.includes(Number.parseInt(input)) && String(input).charAt(0) === '2';

/**
 * Check if a given status code is a 3xx redirect response
 * @public
 *
 * @param {string|number} input - Status code
 *
 * @returns {boolean} True if code is 3xx
 */
status.isRedirect = (input) => status.codes.includes(Number.parseInt(input)) && String(input).charAt(0) === '3';

/**
 * Check if a given status code is a 4xx client error response
 * @public
 *
 * @param {string|number} input - Status code
 *
 * @returns {boolean} True if code is 4xx
 */
status.isClientError = (input) => status.codes.includes(Number.parseInt(input)) && String(input).charAt(0) === '4';

/**
 * Check if a given status code is a 5xx server error response
 * @public
 *
 * @param {string|number} input - Status code
 *
 * @returns {boolean} True if code is 5xx
 */
status.isServerError = (input) => status.codes.includes(Number.parseInt(input)) && String(input).charAt(0) === '5';

/**
 * Check if a given status code is a 4xx or 5xx response
 * @public
 *
 * @param {string|number} input - Status code
 *
 * @returns {boolean} True if code is 4xx or 5xx
 */
status.isError = (input) => status.isClientError(input) || status.isServerError(input);

/**
 * Copy of STATUS_CODES object from Node's http module
 * @type {Object}
 */
status.STATUS_CODES = http.STATUS_CODES; // copy over http.STATUS_CODES object for convenience

/**
 * Array of status codes
 * @type {number[]}
 */
status.codes = addStatusCodesAndMessages(status); // allow for status[404], status['not found'], and status.codes array

status.codes.infoCodes = status.codes.filter((code) => /1/.test(code.toString().charAt(0)));
status.codes.successCodes = status.codes.filter((code) => /2/.test(code.toString().charAt(0)));
status.codes.redirectCodes = status.codes.filter((code) => /3/.test(code.toString().charAt(0)));
status.codes.clientErrorCodes = status.codes.filter((code) => /4/.test(code.toString().charAt(0)));
status.codes.serverErrorCodes = status.codes.filter((code) => /5/.test(code.toString().charAt(0)));
status.codes.errorCodes = status.codes.filter((code) => /4|5/.test(code.toString().charAt(0)));

status.CONTINUE = 100;
status.SWITCHING_PROTOCOLS = 101;
status.PROCESSING = 102;
status.EARLY_HINTS = 103;

status.OK = 200;
status.CREATED = 201;
status.ACCEPTED = 202;
status.NON_AUTHORITATIVE_INFORMATION = 203;
status.NO_CONTENT = 204;
status.RESET_CONTENT = 205;
status.PARTIAL_CONTENT = 206;
status.MULTI_STATUS = 207;
status.ALREADY_REPORTED = 208;
status.IM_USED = 226;

status.MULTIPLE_CHOICES = 300;
status.MOVED_PERMANENTLY = 301;
status.FOUND = 302;
status.SEE_OTHER = 303;
status.NOT_MODIFIED = 304;
status.USE_PROXY = 305;
status.TEMPORARY_REDIRECT = 307;
status.PERMANENT_REDIRECT = 308;

status.BAD_REQUEST = 400;
status.UNAUTHORIZED = 401;
status.PAYMENT_REQUIRED = 402;
status.FORBIDDEN = 403;
status.NOT_FOUND = 404;
status.METHOD_NOT_ALLOWED = 405;
status.NOT_ACCEPTABLE = 406;
status.PROXY_AUTHENTICATION_REQUIRED = 407;
status.REQUEST_TIMEOUT = 408;
status.CONFLICT = 409;
status.GONE = 410;
status.LENGTH_REQUIRED = 411;
status.PRECONDITION_FAILED = 412;
status.PAYLOAD_TOO_LARGE = 413;
status.URI_TOO_LONG = 414;
status.UNSUPPORTED_MEDIA_TYPE = 415;
status.RANGE_NOT_SATISFIABLE = 416;
status.EXPECTATION_FAILED = 417;
status.IM_A_TEAPOT = 418;
status.MISDIRECTED_REQUEST = 421;
status.UNPROCESSABLE_ENTITY = 422;
status.LOCKED = 423;
status.FAILED_DEPENDENCY = 424;
status.UNORDERED_COLLECTION = 425;
status.UPGRADE_REQUIRED = 426;
status.PRECONDITION_REQUIRED = 428;
status.TOO_MANY_REQUESTS = 429;
status.REQUEST_HEADER_FIELDS_TOO_LARGE = 431;
status.UNAVAILABLE_FOR_LEGAL_REASONS = 451;

status.INTERNAL_SERVER_ERROR = 500;
status.NOT_IMPLEMENTED = 501;
status.BAD_GATEWAY = 502;
status.SERVICE_UNAVAILABLE = 503;
status.GATEWAY_TIMEOUT = 504;
status.HTTP_VERSION_NOT_SUPPORTED = 505;
status.VARIANT_ALSO_NEGOTIATES = 506;
status.INSUFFICIENT_STORAGE = 507;
status.LOOP_DETECTED = 508;
status.BANDWIDTH_LIMIT_EXCEEDED = 509;
status.NOT_EXTENDED = 510;
status.NETWORK_AUTHENTICATION_REQUIRED = 511;

export default status;
